src/server/plugins/scan.js

```diff
@@ -45,15 +45,20 @@ module.exports = async (kbnServer, server, config) => {
   });
 
   for (let path of pluginPaths) {
-    let modulePath;
     try {
-      modulePath = require.resolve(path);
-    } catch (e) {
-      warning({ tmpl: 'Skipping non-plugin directory at <%= path %>', path: path });
-      continue;
+      const modulePath = require.resolve(path);
+      await plugins.new(path);
+      debug({ tmpl: 'Found plugin at <%= path %>', path: modulePath });
+    } catch (err) {
```

Comment:
I don't think we should be using try/catch to manage control flow like this, it's just way too heavy handed, and it could result in one truly unexpected exception being treated as a totally normal scenario and then the plugin will end up being loaded under a condition it didn't expect.

Instead, how about moving the bulk of this logic into PluginCollection, with something like this:

```js
async new(path) {
  const hasIndex = await fileExists(join(path, 'index.js'));
  return hasIndex ? this.newFromIndex(path) : this.newFromPackageJson(path);
}

// the old new()
async newFromIndex(path) {
}

async newFromPackageJson(path) {
}
```